Use configure_file() to configure version only

trivialfis opened the pull request:
@hcho3  This keeps most of your clean up, but added back the intrin detection code in order to avoid writing `build_config.h` in include directory.  After this, only version upgrade can change that file.

hcho3 commented:
@trivialfis If the goal is to avoid polluting the include directory, can we apply something like dmlc/dmlc-core#551?

hcho3 commented:
> After this, only version upgrade can change that file

We should formally enforce this, rather than making it an unwritten convention. Perhaps create a separate config step just for the version (`version_config.h`)

trivialfis commented:
@hcho3 Good point.

hcho3 commented:
Let me try incorporating ideas from dmlc/dmlc-core#551 and dmlc/dmlc-core#566 and get back to you.

trivialfis commented:
@hcho3 I removed the `build_config.h` completely and moved those macros into `base.h` since that's where most of our macros defined, also previously `base.h` includes `build_config.h`.

I also added a `version_config.h` that does only configuring version information.

hcho3 commented on include/xgboost/base.h:
```diff
@@ -232,7 +246,4 @@ using XGBoostVersionT = int32_t;
 }  // namespace xgboost
 
-/* Always keep this #include at the bottom of xgboost/base.h */
-#include <xgboost/build_config.h>
```
Can we include `xgboost/version_config.h` here? Otherwise, we'll have to fret about which header gets included first.

trivialfis replied:
@hcho3 I don't think version have any effect to the rest of the macros right?

hcho3 commented on src/common/hist_util.cc:
```diff
@@ -2,14 +2,16 @@
 
+#include "xgboost/base.h"
 #include "../common/common.h"
+#include "./hist_util.h"
```
This re-arrangement of headers can be avoided by including `version_config.h` in `base.h`.

trivialfis replied:
@hcho3 I will remove some re-arrangement of headers.  The `base.h` is included here for mm prefetch , not for version.

hcho3 commented on src/data/data.cc:
```diff
@@ -2,15 +2,16 @@
+#include "sparse_page_writer.h"
+#include "simple_dmatrix.h"
+#include "simple_csr_source.h"
+
```
This re-arrangement of headers can be avoided by including `version_config.h` in `base.h`.

trivialfis replied:
@hcho3 It might be better to include what you use.  The linter checks for this convention for standard headers, but not for others.

hcho3 commented on src/c_api/c_api.cc:
```diff
@@ -16,6 +10,12 @@
+#include "xgboost/learner.h"
+#include "xgboost/c_api.h"
+#include "xgboost/logging.h"
+#include "xgboost/version_config.h"
```
Why is this change necessary?

trivialfis replied:
@hcho3 Just couldn't help myself when adding `version_config.h` here (we added some consistency around `include ""` and `include <>` when exporting c++ header).  I can revert this as unnecessary change.

hcho3 commented:
Note: Should we need to use `configure_file()` for other things, we should try to apply lessons learned from dmlc/dmlc-core#551 and dmlc/dmlc-core#566.

trivialfis commented:
@hcho3 Agreed. until we can remove those makefiles, your method works better.

hcho3 commented:
Will merge once tests pass.

trivialfis commented:
@hcho3 Thanks for the help. ;-)
